### ISSUE TITLE
Support Ollama LLM model pulling

### ISSUE THREAD

**Issue by devtobi** [topgrade-rs/topgrade#988]
## I want to suggest a new step

* Which tool is this about? Where is its repository?
I want topgrade to support `ollama`, a tool for managing LLMs locally, see https://github.com/ollama/ollama.
Topgrade should invoke model pulling to update all locally installed models.

* Which operating systems are supported by this tool?
Windows, macOS, Linux

* What should Topgrade do to figure out if the tool needs to be invoked?
if `ollama` is found in path it should be invoked,  this also has to work when `ollama` was installed through package managers like homebrew.

* Which exact commands should Topgrade run?
I am currently running through a custom command with `ollama list | awk 'NR>1 && !/reviewer/ {system(\"ollama pull \"$1)}'`
This allows to pull all new model files for currently installed models.
I am not sure if this command works OS independent.

* Does it have a `--dry-run` option? i.e., print what should be done and exit
No dry run option available

* Does it need the user to confirm the execution? And does it provide a `--yes`
  option to skip this step?
No user confirmation needed

## More information
To test this a model should be downloaded on the local machine first, e.g. via `ollama run qwen2.5-coder:0.5b`.
When topgrade invokes `ollama` it should show an output similar to
```
pulling manifest
pulling 170370233dd5... 100% ▕████████████████████████████████████████████████████▏ 4.1 GB
pulling 72d6f08a42f6... 100% ▕████████████████████████████████████████████████████▏ 624 MB
pulling 43070e2d4e53... 100% ▕████████████████████████████████████████████████████▏  11 KB
pulling c43332387573... 100% ▕████████████████████████████████████████████████████▏   67 B
pulling ed11eda7790d... 100% ▕████████████████████████████████████████████████████▏   30 B
pulling 7c658f9561e5... 100% ▕████████████████████████████████████████████████████▏  564 B
verifying sha256 digest
writing manifest
success
```

**Comment by SteveLauC** [comment]
I will try implementing it recently.

**Comment by SteveLauC** [comment]
Noticed that there is a tag along with the model name, which is kinda similar to `docker/podman images`, could this tag be a version number? If so, what will `$ ollama pull <model_name>` do? Update it to `latest` or do nothing

```sh
$ ollama list
NAME                ID              SIZE      MODIFIED      
moondream:latest    55fc3abd3867    1.7 GB    5 minutes ago  
```

**Comment by devtobi** [comment]
@SteveLauC If you run the `pull` command without adding a tag, it will always pull the `latest` tagged image besides the other tags you already have on your system. 
However since `ollama list` includes the image tags in the output, my command I mentioned above should preserve only pulling the latest versions of the respective tagged images.
This is desired because the models have different sizes in parameters e.g. `2b, 7b` and are made for different purposes e.g. `instruct, chat`. So always updating to the `latest` tag is not what you want in this case (except if you explicitely installed `latest`via `ollama pull my-model` or `ollama pull my-model:latest`)

**Comment by SteveLauC** [comment]
Thanks for the explanation, I will pull the model with a tag included so that Topgrade won't do anything that users do not want to do.

**Comment by SteveLauC** [comment]
Implemented in #1001, anyone wants to give it a test? I can provide Linux/x86_64 and macOS/arm64 binaries. For the macOS build, I am not sure if it can be executed on other Macs, it seems that the binary needs to be signed.

**Comment by SteveLauC** [comment]
Here is what it looks like on my machine:

```sh
$ ./target/debug/topgrade --only ollama

── 18:51:48 - Sudo ─────────────────────────────────────────────────────────────

── 18:51:48 - Ollama ───────────────────────────────────────────────────────────
Pulling model 'gemma2:2b'
pulling manifest 
pulling 7462734796d6... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏ 1.6 GB                         
pulling e0a42594d802... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏  358 B                         
pulling 097a36493f71... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏ 8.4 KB                         
pulling 2490e7468436... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏   65 B                         
pulling e18ad7af7efb... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏  487 B                         
verifying sha256 digest 
writing manifest 
success 
Pulling model 'moondream:latest'
pulling manifest 
pulling e554c6b9de01... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏ 828 MB                         
pulling 4cc1cb3660d8... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏ 909 MB                         
pulling c71d239df917... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏  11 KB                         
pulling 4b021a3b4b4a... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏   77 B                         
pulling 9468773bdc1f... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏   65 B                         
pulling ba5fbb481ada... 100% ▕██████████████████████████████████████████████████████████████████████████████████████▏  562 B                         
verifying sha256 digest 
writing manifest 
success 

── 18:51:51 - Summary ──────────────────────────────────────────────────────────
Ollama: OK
```

**Comment by devtobi** [comment]
Looks great :) Whats the quickest way for me to test? Should I just checkout your branch and build myself?

**Comment by SteveLauC** [comment]
If you have `rustup` installed, yeah, you can build it yourself:

```sh
$ git clone https://github.com/SteveLauC/topgrade.git
$ cd topgrade
$ cargo build
$ ./target/debug/topgrade --only ollama
```

If not, I can build it for you and upload it here: https://github.com/SteveLauC/topgrade/releases

**Comment by devtobi** [comment]
My local environment is currently not set up for Rust development, so I would be happy to have a provided binary from your side. ;) We will see if it works out with the signing problem. ;)

**Comment by SteveLauC** [comment]
There you go: https://github.com/SteveLauC/topgrade/releases/tag/pr1001

**Comment by devtobi** [comment]
Thanks for sharing. Everything works like a charm with the binary you provided if only using pulled models. 😄 I get the same output as you shared above.

However we have an issue, when models are built locally. Just like with Docker and `Dockerfile`, ollama allows building custom models using `Modelfile`.

The problem now is, that the command `ollama list` also outputs the locally built models and thus topgrade tries to pull them.
This leads to an error like `no manifest file found` and the new "Ollama" step of topgrade exits with a fail.

Unfortunately `ollama list` does not provide a flag to only list pulled model. As far as I saw it ollama saves its models in a specific user directory. Maybe we could find out which models are pulled/created locally this way.

However I am not 100% sure if the effort is justified. Most users most likely only use downloaded models. So this might be kind of an edge case I described here.

As already mentioned above the pulling of models works without any issues, even though topgrade reports a failure due trying to pull locally created models.

We still need to make sure models are further pulled down after a "pull error" occured. I don't know if topgrade stops the executing immediatly after an error occurred or not. In my case the failing model pull was executed last, so I couldn't reproduce this edge case.

**Comment by adben** [comment]
Acutally I have some question about my approach....

I'm currently using custom commands in my topgrade.toml to update Ollama and its models:

```toml
[commands]
"Update Ollama" = "bash -c 'current=$(ollama --version 2>&1 | grep -oP \"\\d+\\.\\d+\\.\\d+\"); latest=$(curl -s https://api.github.com/repos/ollama/ollama/releases/latest | grep -oP \"\\\"tag_name\\\":\\s*\\\"v\\K[0-9.]+\"); if [[ \"$current\" != \"$latest\" ]]; then echo \"Upgrading Ollama from $current to $latest\"; curl -fsSL https://ollama.com/install.sh | sh; else echo \"Ollama $current is already the latest version\"; fi'"
"Update Ollama Models" = "ollama list 2>/dev/null | awk 'NR>1 {print $1}' | xargs -I {} sh -c 'echo \"Updating model: {}\"; ollama pull {} || echo \"Failed to update {}\"; echo \"--\"' && echo \"All models updated.\""
```

While this works, I'm looking for advice on how to integrate Ollama updates more natively with Topgrade. 

## Questions for the Community

1. Would a dedicated `[ollama]` section in Topgrade make sense for others too?
2. Are there cleaner approaches than my current bash commands?
3. Should this be implemented as a Topgrade plugin instead? 
4. Has anyone else implemented something similar with better practices?

I'd appreciate any suggestions to improve this approach or make it more maintainable and reliable.

Thank you!

**Comment by Beeblebrox21** [comment]
> While this works, I'm looking for advice on how to integrate Ollama updates more natively with Topgrade.
> ## Questions for the Community
> 
>     1. Would a dedicated `[ollama]` section in Topgrade make sense for others too?

For me this makes definitely sence.

>     2. Are there cleaner approaches than my current bash commands?

Works for me.
 
>     3. Should this be implemented as a Topgrade plugin instead?

This might be an elegant solution, it keeps the main tool small and not every user uses LLMs
 

just my 2¢ ...

**Comment by GideonBear** [comment]
> 3. Should this be implemented as a Topgrade plugin instead?

No, we don't do plugins; instead we detect if the user has a tool installed, and update automatically; or, if it's not always (or mostly not) desired to update by default, make it opt-in.

There is an open PR, however it is blocked on a solution for https://github.com/topgrade-rs/topgrade/issues/988#issuecomment-2532175570.

**Comment by devtobi** [comment]
@SteveLauC Do you still have interest in getting this ready for PR? As stated in https://github.com/topgrade-rs/topgrade/issues/988#issuecomment-2532175570 the problem with the "pull error" for local models should be kind of an edge case. So in my opinion we could ignore that. @GideonBear What do you think?

However we should still make sure pulling a unavailable model (which could also be the case for externally downloaded models once the repository or file gets deleted) does not stop the execution.

**Comment by GideonBear** [comment]
Hi @devtobi! @SteveLauC is probably not going to work on the PR anymore, I'll hijack or recreate the PR to rebase it when we've worked out what to do.

I don't think the current strat (ignoring the problem and letting Topgrade crash when you have local models) is good enough. However, I agree that trying to figure out which models are local is, for now at least, not necessary.

What do you think about simply ignoring any errors that `ollama pull` throws? Something like:
```console
$ topgrade --only ollama
Step ollama
INFO Found models ['a', 'b']
INFO Pulling model 'a'
ollama: Pulling 100%
INFO Pulling model 'b'
ollama: no manifest file found
WARN Command `ollama pull b` failed! This is expected for local models. Please check the ollama error message for more information.
```

**Comment by devtobi** [comment]
> Hi [@devtobi](https://github.com/devtobi)! [@SteveLauC](https://github.com/SteveLauC) is probably not going to work on the PR anymore, I'll hijack or recreate the PR to rebase it when we've worked out what to do.
> 
> I don't think the current strat (ignoring the problem and letting Topgrade crash when you have local models) is good enough. However, I agree that trying to figure out which models are local is, for now at least, not necessary.
> 
> What do you think about simply ignoring any errors that `ollama pull` throws? Something like:
> 
> $ topgrade --only ollama
> Step ollama
> INFO Found models ['a', 'b']
> INFO Pulling model 'a'
> ollama: Pulling 100%
> INFO Pulling model 'b'
> ollama: no manifest file found
> WARN Command `ollama pull b` failed! This is expected for local models. Please check the ollama error message for more information.

That's what I tried to suggest. As long as we make sure other models still get pulled when a pull error occurs I am more than happy. :) However note pull errors can happen:
- for local models
- for remote models that are no longer available
So I'd adjust the error message accordingly to explain both cases.

However LGTM so far!

**Comment by devtobi** [comment]
@GideonBear just took a look back into the discussion of this discussion. Any updates on the status of a PR? You already had kind of a good idea in your last comment. 😉

**Comment by GideonBear** [comment]
It's on my todo list, haven't had the time for it yet.

**Comment by uwuclxdy** [comment]
I've been working on this, @GideonBear you can assign me
